parlai/core/teachers.py

```diff
@@ -927,6 +933,8 @@ def get(self):
                 if self.entry_idx == 0:
                     self.cur_episode = self.next_episode
                     self.next_episode = next(self.data)
+                if self.cur_episode == -1:
```

Comment:
all this code can be simplified to this:

```python
if self.cur_episode is None:
    # first go around, always read off the episode
    self.cur_episode = next(self.data)
if self.cur_episode == -1:
    # we're done here, but try to prep for next time
    return {'episode_done': True}, True
entry = self.cur_episode[self.entry_idx]
table = self.build_table(entry)
episode_done = self.entry_idx == len(self.cur_episode) - 1
table['episode_done'] = episode_done
if episode_done:
    self.cur_episode = next(self.data)
    self.entry_idx = 0
else:
    self.entry_idx += 1
return table, self.cur_episode == -1
raise ValueError
```

If you think this is better than the current next_episode madness, then we should update this change.

Reply:
this whole `-1` thing is so bizarre, could we have this like defined as a constant or a type? like `self.cur_episode == END_OF_EPOCH` or something?

Reply:
assuming the `raise ValueError` here is a typo?

Reply:
i think this solution is better than the `next_episode` madness btw

Reply:
Okay, I'll propose the new change.

Reply:
agree with @klshuster on the -1 and that this is better. i also find `build_table` to be a confusing name for just creating the next observation/action dict.

Reply:
also do we need that `_lock()` still

Reply:
Ya I think we probably do.